### DISCUSSION TITLE
Diacritic characters not recognized

### DISCUSSION THREAD

**Discussion by Coooz** [nextgenhealthcare/connect#5457] [Q&A]
Hi,

I noticed with a customer of ours that a Mirth Connect channel does not receive diacritic characters correctly. This is an example of a  message in Notepad:
![image](https://user-images.githubusercontent.com/88899312/194318378-10421750-548c-4a07-b198-70c49f078397.png)

And this is what the message looks like using the Send Message functionality in Mirth Connect:
![image](https://user-images.githubusercontent.com/88899312/194317659-2030cf56-0178-4c65-b7a2-61634a493938.png)
The message in Notepad shows that the red square indicates an é, and that the blue square indicates an ö. If I copypaste the message from Notepad in the Send Message form, then the characters show up correctly. But obviously I can't use this in practice. Actually the messages are picked up by a TCP Listener channel.

I noticed that on my local machine, in a different environment, this issue doesn't show up. So, what does this behavior depend upon?

How can I get the characters right again? I'm using Mirth Connect 3.12.
Thanks!

**Comment by ab-20-m** [comment]
Couple questions:
1. Is the TCP Listener configured to use the 'Default Encoding' or an explicit character encoding?
2. What OS is the Mirth instance running on?
3. What version of Java is the Mirth instance using?

**Reply by Coooz** [reply]
Hi @ab-20-m ,
1. It was Default. I changed it to UTF-8 and tried again, but no difference. I will try a couple of other encodings now.
2. Windows Server 2012 R2
3. 
![image](https://user-images.githubusercontent.com/88899312/194327784-4fe3b208-8e6c-4ed4-81aa-52f53346c1e7.png)

**Reply by pacmano1** [reply]
Seems fine on MacOS, mirth v4.0.1 though

![image](https://user-images.githubusercontent.com/44065187/194341986-7827f566-3aef-4b79-9f64-c42b191fba21.png)

**Reply by pacmano1** [reply]
3.12 on windows 10:

![image](https://user-images.githubusercontent.com/44065187/194344071-f41ebf55-83cd-47d1-b7f9-32ee1168bba2.png)

**Reply by pacmano1** [reply]
And not that it matter but an HL7 message for 3.12 on windows 10:

![image](https://user-images.githubusercontent.com/44065187/194344642-e44a4402-aeaa-44d9-957d-5562af333ce3.png)

**Reply by Coooz** [reply]
@pacmano1 
Yes, well, it looks fine on my local Windows 10 machine as well, but I need to get it fixed in the customer's environment.

Can it be a java issue? I use OpenJDK 11 and many of our customers do - even this customer has OpenJDK 11 in a different environment used by Mirth Connect, and that seems okay. I have no access there but neither do I hear that the problem pops up on that environment.
The java version used by the Mirth Connect instance where things go wrong is java 1.8.0.

**Reply by pacmano1** [reply]
Other than trying all OS/Encoding/Java/Java Vendor combinations on your own, not sure how to sort that out.

**Reply by siddharthsharma9258** [reply]
Bit of a long shot but a python script can help you to detect the system default encoding. If you try to read a file in a small python script with no specified encoding, it will throw an exception as below

![image](https://user-images.githubusercontent.com/20477452/194372285-5f1192ae-7cd1-4e50-89b4-07d0e39e2a98.png)

the yellow highlight is the encoding scheme that the system is using.

**Reply by Coooz** [reply]
@siddharthsharma9258 Thanks for your suggestion - but the system with the issue appeared to have the same default encoding as my local machine, which does not show this behavior in Mirth Connect.
Well, at least we've ruled out that the system's encoding is the culprit.

**Reply by Coooz** [reply]
I found out that it is not a java issue, as I earlier suggested, since the behavior appears to happen in the customer's other environment as well. Java being the culprit would have been strange of course, but _something_ must be causing this...

**Reply by ab-20-m** [reply]
@Coooz to be clear, the scenario is when you use the Open Text File button in the Send Message dialog and point at a file, you get the squares instead the expected characters? I was able to replicate that scenario on Windows Server 2019, Mirth 3.12.0 and Java 8u144 or newer.

As best I can tell there might be something going in in how JEdit was implemented possibly in combination with being on Windows. I'm really not sure why Copy-Paste would work.

**Reply by Coooz** [reply]
@ab-20-m Yes, that is the scenario! At least I'm not alone any longer... 😄

I think I'm making some progress by now, since the NextGen Connect 312 user guide mentions about the Send message functionality: "When reading a file, you can choose to open it as a 'Text' file, _which will use the default Java charset encoding_." (my italics)
So now I try to change the default Java charset encoding in the channel's Deploy script, before any messages arrive. Logically speaking, that should do the trick. 
Still fumbling somewhat with importing and calling the needed Java packages though... that's new territory to me.

**Reply by siddharthsharma9258** [reply]
> I'm really not sure why Copy-Paste would work.

Maybe it is the Notepad native feature that strips all formatting in the text that is helping here. Every thread on google says Notepad does that, but none of it actually explains why it does so.

**Reply by Coooz** [reply]
"So now I try to change the default Java charset encoding in the channel's Deploy script, before any messages arrive. Logically speaking, that should do the trick."
But it doesn't: the default Java charset encoding appears to be Cp1252 in both environments - one with and one without the behavior I mean. Bummer. I'm rapidly running out of possibilities here. Anyone? 🥺

**Reply by ab-20-m** [reply]
The documentation is outdated/incorrect. When you use Open Text File dialog, the underlying code actually ALWAYS reads the files as UTF-8 encoded. If you want to load files with Open Text File, the files need to be saved as UTF-8 encoded without a BOM. Otherwise, the BOM will come in similar to what you are seeing currently.

**Reply by Coooz** [reply]
@ab-20-m Yes, you are right - the (official!) documentation is incorrect; files are read as UTF-8. The issue disappears when sending UTF-8 encoded messages. Thank you!

Okay, that was part 1. 😉

I cannot be sure that incoming messages will be in UTF-8 encoding - in fact I am pretty sure that most of them are not. 
Therefore: can I use the Preprocessor script to get the encoding of an incoming message and change it to UTF-8 if it isn't already? Would that work - and can someone point me in the right direction as to what that script would look like?

Thanks once again...

**Reply by siddharthsharma9258** [reply]
Your problem sounds a lot like this thread : https://stackoverflow.com/questions/11497902/how-to-check-the-charset-of-string-in-java

Even if you could pull something together, I doubt that you could make it work inside the preprocessor script. I have never seen external libs being used in preprocessor cos it will add to the delay in receiving.

It's highly probable that the encoding scheme is ISO-8859-1.

**Reply by ab-20-m** [reply]
The MSH-18 field is meant for describing the encoding(s) of a message. The field is optional and many systems do not set the field, in which case the standard is to assume ASCII 7-bit unless the sending system has documented the encoding in an implementation agreement or other such document.

Assuming the sending system did set the MSH-18 field, HAPI has a [CharSetUtil](https://github.com/hapifhir/hapi-hl7v2/blob/master/hapi-base/src/main/java/ca/uhn/hl7v2/llp/CharSetUtil.java) class that handles trying to detect the encoding based on MSH-18. However the method you would want access to, `static Charset checkCharset(byte[] message, Charset defaultCharset)`, is package-private. You would have to reimplement that code and switch the TCP Listener from Text to Binary so you can access the raw bytes from the network.

**Reply by Coooz** [reply]
@ab-20-m Yes, I know about MSH.18. The sending system does not set it, and it cannot be set to UTF-8.

**Comment by Coooz** [comment]
It is fixed on the side of the sending system - the encoding of the messages sent is made configurable. It looks good now.
Would've been nice if it could have been fixed on the Mirth Connect side though...
